File: java/code/34.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int start = 0;
        int end = nums.length - 1;

        int left = -1, right = -1;
        while(start <= end){
            int middle = (start + end) / 2;
            if (nums[middle] == target && (middle == 0 || nums[middle-1] != target)){
                left = middle;
                break;
            }
            if (nums[middle] < target) start = middle + 1;
            else end = middle - 1;
        }
        start = 0;
        end = nums.length - 1;
        while(start <= end){
            int middle = (start + end) / 2;
            if (nums[middle] == target && (middle == nums.length - 1 || nums[middle+1] != target)){
                right = middle;
                break;
            }
            if (nums[middle] > target) end = middle - 1;
            else start = middle + 1;
        }
        int ans[] = {left, right};
        return ans;
    }
}

File: C++/code/34.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int n = nums.size();
        int left = 0, right = n - 1, ansleft = -1, ansright = -1;
        vector<int> ans{-1, -1};
        if (nums.empty()) return ans;
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] == target) right = mid;
            else if (nums[mid] > target) right = mid - 1;
            else left = mid + 1;
        }
        ansleft = left;
        left = 0, right = n - 1;
        while (left < right) {
            int mid = left + (right - left + 1) / 2;
            if (nums[mid] == target) left = mid;
            else if (nums[mid] > target) right = mid - 1;
            else left = mid + 1;
        }
        ansright = left;
        if (nums[ansleft] == target) {
            ans[0] = ansleft;
            ans[1] = ansright;
        }
        return ans;
    }
};